applications/SystemIdentificationApplication/custom_sensors/measurement_residual_p_norm_response_function.cpp

```diff
@@ -0,0 +1,307 @@
+        const ProcessInfo& rProcessInfo)
+    {
+        rSensor.CalculatePartialSensitivity(rEntity, rVariable, rSensitivityMatrix, rSensitivityGradient, rProcessInfo);
+    }
+};
+
+} // namespace MeasurementResidualPNormResponseFunctionUtilities
+
+MeasurementResidualPNormResponseFunction::MeasurementResidualPNormResponseFunction(const double PCoefficient)
+    : mPCoefficient(PCoefficient)
+{
+    mResponseGradientList.resize(ParallelUtilities::GetNumThreads());
+}
+
+void MeasurementResidualPNormResponseFunction::AddSensor(Sensor::Pointer pSensor)
+{
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF_NOT(pSensor->Has(SENSOR_MEASURED_VALUE))
+        << pSensor->GetName() << " does not have the SENSOR_MEASURED_VALUE defined.\n";
+
+    mpSensorsList.push_back(pSensor);
+
+    KRATOS_CATCH("");
+}
+
+void MeasurementResidualPNormResponseFunction::Clear()
+{
+    mpSensorsList.clear();
+}
+
+std::vector<Sensor::Pointer>& MeasurementResidualPNormResponseFunction::GetSensorsList()
+{
+    return mpSensorsList;
+}
+
+void MeasurementResidualPNormResponseFunction::Initialize()
+{
+    for (auto& p_sensor : mpSensorsList) {
+        p_sensor->Initialize();
+    }
+}
+
+void MeasurementResidualPNormResponseFunction::InitializeSolutionStep()
+{
+    for (auto& p_sensor : mpSensorsList) {
+        p_sensor->InitializeSolutionStep();
+    }
+}
+
+void MeasurementResidualPNormResponseFunction::FinalizeSolutionStep()
+{
+    for (auto& p_sensor : mpSensorsList) {
+        p_sensor->FinalizeSolutionStep();
+    }
+}
+
+void MeasurementResidualPNormResponseFunction::Finalize()
+{
+
+}
+
+double MeasurementResidualPNormResponseFunction::CalculateValue(ModelPart& rModelPart)
+{
+    KRATOS_TRY
+
+    double value = 0.0;
+    for (auto& p_sensor : mpSensorsList) {
+        const double sensor_value = p_sensor->CalculateValue(rModelPart);
+        p_sensor->SetSensorValue(sensor_value);
+        const double current_sensor_error_square  = std::pow(sensor_value - p_sensor->GetValue(SENSOR_MEASURED_VALUE), 2);
+        p_sensor->SetValue(SENSOR_ERROR, current_sensor_error_square);
+        value += std::pow(p_sensor->GetValue(SENSOR_ERROR), mPCoefficient);
+    }
+
+    return std::pow(value, 1 / mPCoefficient);
+
+    KRATOS_CATCH("");
+}
+
+template<class TCalculationType, class... TArgs>
+void MeasurementResidualPNormResponseFunction::CalculateDerivative(
+    Vector& rResponseGradient,
+    const Matrix& rResidualGradient,
+    TArgs&&... rArgs)
+{
+    KRATOS_TRY
+
+    if (rResponseGradient.size() != rResidualGradient.size1()) {
+        rResponseGradient.resize(rResidualGradient.size1(), false);
+    }
+
+    rResponseGradient.clear();
+
+    auto& local_sensor_response_gradient = mResponseGradientList[OpenMPUtils::ThisThread()];
+    double temp = 0.0;
+    for (auto& p_sensor : mpSensorsList) {
+        temp += ( std::pow( p_sensor->GetValue(SENSOR_ERROR), mPCoefficient ) );
+    }
+    double c1 = 1 / mPCoefficient * std::pow( temp, 1/mPCoefficient - 1 );
```

Comment:
```suggestion
    const double c1 = 1 / mPCoefficient * std::pow( temp, 1/mPCoefficient - 1 );
```

Reply:
This is what I said which is dangerous in `p-norm`. Because, our `SENSOR_ERROR` usually tends to zero (not now, may be in future ;)), then `std::pow( temp, 1/mPCoefficient - 1 )` will be infinity [`mPCoefficient > 1.0`]. Need to be careful here.